### ISSUE TITLE
[Bug]: Variable-length DP/EP requests can hang when MC2 skips ACLGraph mode sync

### ISSUE THREAD

**Issue by QwertyJack** [vllm-project/vllm-ascend#13826]
### Environment

- Hardware: Ascend A2, 8 nodes (4 prefill + 4 decode)
- Model: DeepSeek-V4-Flash-0731 W8A8 with DSpark speculative decoding
- Decode topology: DP32 / EP32 / TP1, PD KV consumer
- vLLM: `v0.25.1` (`752a3a504485790a2e8491cacbb35c137339ad34`)
- vLLM Ascend: `releases/v0.25.1rc` (`50e0ce60849a7b6fe6f94e1de085b6fc06e7006d`)
- Target compilation mode: `CUDAGraphMode.FULL_DECODE_ONLY`
- DSpark drafter: eager
- `max_num_batched_tokens=256`, `max_num_seqs=32`
- Async scheduling and expert parallelism enabled

The same missing synchronization is still present on current `main` at
`d6e8f1ad4bd67cc48a349cd7cf8efab98eecad13`.

### Problem

Fixed-length requests complete, but variable-length requests can leave the decode
service permanently stuck:

- Fixed 16K input / 1K output: completes.
- Uniform 1K-2K input / 1K-2K output, batch size 32, 128 requests: can hang.
- All NPUs remain busy while only a subset of DP ranks still report running requests.
- The service does not recover after the client run ends.

This is most visible at the variable-length tail, when some DP ranks become idle
before others.

### Root cause

`NPUModelRunner._sync_metadata_across_dp()` returns early when
`should_skip_allreduce_across_dp_group()` is true:

```python
if should_skip_allreduce_across_dp_group(self.vllm_config, is_draft_model):
    num_tokens_after_padding = torch.tensor(
        [num_tokens] * self.dp_size, device="cpu", dtype=torch.int32
    )
    return num_tokens, num_tokens_after_padding, cudagraph_mode
```

Skipping token-count synchronization is valid for the MC2 uneven-token path, but
the same early return also skips synchronization of the ACLGraph runtime mode.

At the variable-length tail this can produce:

- busy rank: `CUDAGraphMode.FULL`
- idle/dummy rank: `CUDAGraphMode.NONE`

The ranks then submit graph and eager work into the same EP collective. On the
reproduced A2 configuration this leaves the HCCL collective waiting indefinitely.

### Isolation and validation

A DSA metadata padding patch alone did not prevent the hang.

Making only the decode target globally eager isolated the issue to the target
ACLGraph path. The DSpark drafter was already eager.

| Configuration | Run 1 | Run 2 |
| --- | --- | --- |
| Target ACLGraph, without mode-sync fix | completed | hung |
| Target globally eager | 128/128 | 128/128 |
| Target ACLGraph, with mode-sync fix | 128/128 | 128/128 |

For both fixed runs:

- Failed requests: 0
- Service errors: 0
- All 32 decode endpoints remained healthy
- All ranks returned to `Running=0`, `Waiting=0`

The two ACLGraph fixed runs completed in 126.6 s and 115.7 s.

### Proposed fix

Keep the MC2 per-rank token counts, but synchronize only the runtime graph mode
for the main MoE model when ACLGraph is enabled:

```python
mode_tensor = torch.tensor(cudagraph_mode.value, device="cpu", dtype=torch.int32)
dist.all_reduce(
    mode_tensor,
    op=dist.ReduceOp.MIN,
    group=get_dp_group().cpu_group,
)
cudagraph_mode = CUDAGraphMode(mode_tensor.item())
```

Taking the minimum preserves the existing rule that if any rank requires eager
execution (`NONE`), every DP rank uses eager execution for that step.

The fix should not affect the dense/eager DSpark drafter and should not restore
token padding.

### A3 scope

Current `main` can also select MC2 or FUSED_MC2 on A3, so the same code path is
potentially affected when hierarchy communication is disabled. This report only
claims an A2 hardware reproduction; A3 still requires real-hardware validation.
